pcdsdevices/tests/test_tpr.py

```diff
@@ -22,3 +22,29 @@ def test_enable(fake_trigger):
 @pytest.mark.timeout(5)
 def test_disconnected_trigger():
     TprTrigger('TST', channel=1, timing_mode=TimingMode.LCLS1, name='test')
+
+
+def put_equals_setpoint(mds, setpoint, ns_time):
+    mds.put(ns_time)
+    return setpoint.get() == ns_time
+
+
+def mds_get(mds, tick_signal, tap_signal, num_ticks, num_taps):
+    tick_signal.put(num_ticks)
+    tap_signal.put(num_taps)
+    return mds.get() == (TPR_TICK_NS * num_ticks + TPR_TAP_NS * num_taps)
+
+
+def test_ns_delay(fake_trigger):
```

Comment:
```suggestion
@pytest.mark.timeout(5)
def test_ns_delay(fake_trigger):
```
Adding a timeout here makes this run consistently locally.  We do this a lot in the pcdsdevices test suite, and you'll probably need it for the `test_width` test as well

Reply:
With the timeout added, it still fails locally sometimes and failed the the github job. This is the error message
`E       AssertionError: assert False
E        +  where False = put_equals_setpoint(MultiDerivedSignal(name='trig_a_ns_delay', parent='trig_a', value=None, timestamp=1713216587.5022318), FakeEpicsSignal(name='trig_a_delay_setpoint', parent='trig_a', value=0, timestamp=1713216587.5021243), 100)
E        +    where MultiDerivedSignal(name='trig_a_ns_delay', parent='trig_a', value=None, timestamp=1713216587.5022318) = FakeTprTrigger(TST:TPR, name=trig_a).ns_delay
E        +    and   FakeEpicsSignal(name='trig_a_delay_setpoint', parent='trig_a', value=0, timestamp=1713216587.5021243) = FakeTprTrigger(TST:TPR, name=trig_a).delay_setpoint

test_tpr.py:40: AssertionError
`
I don't know why a timeout would help since I'm not hanging on anything. Sleeping for two seconds after the put in put_equals_setpoint (between lines 28 and 29) makes it pass consistently locally but I imagine there there's a better way to ensure that you get the setpoint pv after it's been written to by the MultiDerivedSignal. I tried using code from test_multi_derived_rw_sub and wait_until_value in test_signal, but I never entered the callback after subscribing for reasons I don;'t understand. That test and mine are both using MultiDerivedSignals.

Reply:
Sorry, I assumed here that we just needed to wait for the callbacks to resolve.  I guess the problem here is that the signals already have values, and we need them to update.  It should be easy enough to just poll until the condition is true, given a timeout:

```python
def wait_until(condition, interval=0.1, timeout=1, *args):
  start = time.time()
  while not condition(*args) and time.time() - start < timeout:
    time.sleep(interval)
```

where `condition` is a function wrapping your assert statement or something similar.

Reply:
This time I can get it very consistently passing locally for me, previously on heavy repeats (100) I was getting failures

Reply:
I ended up using Status.wait for the put return along with a timeout on the tests. It works and feels cleaner than spinning or a callback in this case